File: opcua-webservice-spring/core/src/main/java/org/opcfoundation/webserver/addressspace/nodemanager/NodeManagerReactiveObject.java
package org.opcfoundation.webserver.addressspace.nodemanager;

import org.eclipse.milo.opcua.stack.core.types.builtin.*;
import org.eclipse.milo.opcua.stack.core.types.enumerated.*;
import org.eclipse.milo.opcua.stack.core.types.structured.*;
import org.opcfoundation.webapi.service.types.*;
import org.opcfoundation.webserver.service.transactions.base.*;
import org.springframework.lang.Nullable;
import org.opcfoundation.webserver.addressspace.reactiveobject.UaReactiveObjectTransactionManager;
import org.opcfoundation.webserver.addressspace.nodes.UaNode;
import org.opcfoundation.webserver.addressspace.nodes.UaObject;
import org.opcfoundation.webserver.addressspace.reactiveobject.UaReactiveObjectType;
import org.opcfoundation.webserver.types.common.UaBrowseAdditionalInfo;
import org.opcfoundation.webserver.types.common.UaObjectIdentifier;

import java.util.*;

public class NodeManagerReactiveObject extends NodeManager {

    public NodeManagerReactiveObject(String namespaceUri)
    {
        super(namespaceUri);
    }

    public final @Nullable UaReactiveObjectType findObjectType(UaObjectIdentifier objectId)
    {
        NodeId objectTypeId = null;

        if (null != objectId.getTypeId())
        {
            objectTypeId = NodeId.parseOrNull(objectId.getTypeId());
        } else {
            if (null == objectId.getInstanceDeclId()) return null;

            NodeId instanceDeclId = NodeId.parseOrNull(objectId.getInstanceDeclId());
            if (null == instanceDeclId) return null;

            UaNode instanceDeclNode = getNode(instanceDeclId);
            if (null == instanceDeclNode || NodeClass.Object != instanceDeclNode.nodeClass()) return null;

            objectTypeId = ((UaObject)instanceDeclNode).typeDefinition().nodeId();
        }

        if (null == objectTypeId || objectTypeId.getType() != IdType.String) return null;

        UaNode node = getNode(objectTypeId);
        if (null == node || NodeClass.ObjectType != node.nodeClass()) return null;

        return (UaReactiveObjectType)node;
    }

    public final @Nullable UaObject findInstanceDeclaration(UaObjectIdentifier objectId)
    {
        if (null == objectId.getInstanceDeclId()) return null;
        NodeId instanceDeclarationId = NodeId.parseOrNull(objectId.getInstanceDeclId());
        if (null == instanceDeclarationId) return null;

        UaNode instanceDeclaration = getNode(instanceDeclarationId);
        if (null == instanceDeclaration || instanceDeclaration.nodeClass() != NodeClass.Object) return null;
        return (UaObject) instanceDeclaration;
    }

    @Override
    public final UaBrowseTransaction getBrowseTransaction(
            ServiceContext context,
            BrowseDescription nodeToBrowse,
            UaBrowseAdditionalInfo additionalInfo,
            int handleId)
    {
        return new UaReactiveObjectTransactionManager(this).getBrowseTransaction(
                context,
                nodeToBrowse,
                additionalInfo,
                handleId);
    }

    @Override
    public final List<UaReadTransaction> getReadTransactions(
            ReadContext context,
            List<Integer> handleIds)
    {
        return new UaReactiveObjectTransactionManager(this).getReadTransactions(context, handleIds);
    }

    @Override
    public final List<UaWriteTransaction> getWriteTransactions(
            WriteContext context,
            List<Integer> handleIds)
    {
        return new UaReactiveObjectTransactionManager(this).getWriteTransactions(context, handleIds);
    }

    @Override
    public final UaMethodCallTransaction getMethodCallTransaction(
            CallContext context,
            int handleId)
    {
        return new UaReactiveObjectTransactionManager(this).getMethodCallTransaction(context, handleId);
    }

    @Override
    public final UaHistoryReadTransaction getHistoryReadTransaction(
            HistoryReadContext context,
            int handleId)
    {
        return new UaReactiveObjectTransactionManager(this).getHistoryReadTransaction(context, handleId);
    }
}

File: opcua-webservice-node/source/src/server/addressspace/nodemanager/NodeManagerReactiveObject.ts
import { NodeClass } from "opcua-webapi";
import { parseUaNodeIdOrNull, UaBrowseDescription, UaNodeId } from "opcua-webapi-ts";
import {
    CallContext,
    HistoryReadContext,
    ReadContext,
    ServiceContext,
    UaBrowseAdditionalInfo,
    UaObjectIdentifier,
    WriteContext,
} from "../../types";
import {
    UaBrowseTransaction,
    UaHistoryReadTransaction,
    UaMethodCallTransaction,
    UaReadTransaction,
    UaWriteTransaction,
} from "../../service/transactions";
import { UaReactiveObjectTransactionManager } from "../reactiveobject/UaReactiveObjectTransactionManager";
import { UaObject } from "../nodes";
import { UaReactiveObjectType } from "../reactiveobject/UaReactiveObjectType";
import { NodeManager } from "./NodeManager";

export class NodeManagerReactiveObject extends NodeManager {
    constructor(namespaceUri: string) {
        super(namespaceUri);
    }

    findObjectType(objectId: UaObjectIdentifier): UaReactiveObjectType | null {
        let typeId : UaNodeId | null = null;

        if (objectId.typeId !== null) {
            typeId = parseUaNodeIdOrNull(objectId.typeId);
        } else if (objectId.instanceDeclId !== null) {
            let instanceDeclNodeId = parseUaNodeIdOrNull(objectId.instanceDeclId);
            if (instanceDeclNodeId === null) return null;

            const instanceDeclNode = this.getNode(instanceDeclNodeId);
            if (instanceDeclNode === null || instanceDeclNode.nodeClass !== NodeClass.Object) return null;

            const typeDefinitionId = (instanceDeclNode as UaObject).typeDefinition.nodeId;
            typeId = typeDefinitionId;            
        }

        if (typeId === null) return null;
                
        let typeNode = this.getNode(typeId);
        if (typeNode === null || typeNode.nodeClass !== NodeClass.ObjectType) return null;

        return typeNode as unknown as UaReactiveObjectType;
    }

    findInstanceDeclaration(objectId: UaObjectIdentifier): UaObject | null {
        if (objectId.instanceDeclId === null) return null;
        
        let nodeId = parseUaNodeIdOrNull(objectId.instanceDeclId);
        if (nodeId === null) return null;
        
        const instanceDeclaration = this.getNode(nodeId);
        if (instanceDeclaration === null || instanceDeclaration.nodeClass !== NodeClass.Object) {
            return null;
        }

        return instanceDeclaration as UaObject;
    }

    getBrowseTransaction(
        context: ServiceContext,
        nodeToBrowse: UaBrowseDescription,
        additionalInfo: UaBrowseAdditionalInfo,
        handleId: number,
    ): UaBrowseTransaction {
        return new UaReactiveObjectTransactionManager(this).getBrowseTransaction(
            context,
            nodeToBrowse,
            additionalInfo,
            handleId,
        );
    }

    getReadTransactions(
        context: ReadContext,
        handleIds: Array<number>,
    ): Array<UaReadTransaction> {
        return new UaReactiveObjectTransactionManager(this).getReadTransactions(context, handleIds);
    }

    getWriteTransactions(
        context: WriteContext,
        handleIds: Array<number>,
    ): Array<UaWriteTransaction> {
        return new UaReactiveObjectTransactionManager(this).getWriteTransactions(context, handleIds);
    }

    getMethodCallTransaction(
        context: CallContext,
        handleId: number,
    ): UaMethodCallTransaction {
        return new UaReactiveObjectTransactionManager(this).getMethodCallTransaction(context, handleId);
    }

    getHistoryReadTransaction(
        context: HistoryReadContext,
        handleId: number,
    ): UaHistoryReadTransaction {
        return new UaReactiveObjectTransactionManager(this).getHistoryReadTransaction(context, handleId);
    }
}
